### ISSUE TITLE
Automatically sync at cron time

### ISSUE THREAD

**Issue by paglias** [HabitRPG/habitica#4518]
Steps:
- [ ] Calculate next cron time
- [ ] Use $interval
- [ ] Reschedule $interval after CDS changed
- [ ] Reschedule $interval after manual sync or change in `dayStart` or `lastCron` after sync
##

**Comment by Alys** [comment]
Is the idea that this would force a cron to happen at the CDS time even if the user doesn't log in?

**Comment by paglias** [comment]
No, it's just in case the user keeps habit open through cron time, in which
case it sync so that it gets cron results to the browser
Il 14/gen/2015 23:43 "Alice Harris" notifications@github.com ha scritto:

**Comment by Alys** [comment]
So if the HabitRPG website was left open overnight in a browser on a running (not hibernating) computer with an active internet connection, cron would occur at the CDS time?

If HabitRPG had been left open but the computer was hibernating overnight or the internet connection was turned off, would cron occur as soon as the computer or connection was woken up the next morning, even if the user didn't interact with the website?

If the app was running on a phone that was left on overnight, would that also cause cron to occur at CDS?

Sorry about all the questions, it's just that we've been trying to make it very clear to users that cron runs only when they interact with the site, so if that's going to change, we need to understand it well enough to explain it to them. :)

**Comment by deilann** [comment]
It's more when it's left open that when the user ticks off a task (making
cron happen) it forces a sync, yes?

**Comment by Alys** [comment]
But that was already happening wasn't it? A sync occurs every time a task is ticked off?

**Comment by deilann** [comment]
Other things can force cron, though, that don't force a sync.

**Comment by Alys** [comment]
Basically, @paglias, we just need to know how cron behaviour will change so we can describe it to the users. :)

**Comment by paglias** [comment]
@Alys @deilann so what I want to make is that if you leave habit open trough cron time you might see things as they still have to reset but that's not exactly true because the first action you take will trigger cron so what you see on the site is not what you _should_ see. I understand that cron is not meant to run automatically but I want to avoid a situation where while cron has yet to run anything you will do will trigger it possibly causing some unwanted effect (ie if you un-check a task that has to be resetted or check a task that the following day is not set to be done).

It won't cause bugs but users might be confused.

If we don't want to automatically run cron we might show a popup saying:

> Hey, cron time has just passed so we need to fetch the updated data from the server. Clicking ok will trigger a sync

Or something like that, what do you think?

**Comment by deilann** [comment]
no, no, I definitely want it to trigger cron... it should!

**Comment by paglias** [comment]
@Alys as for the `what happens when the computer is suspended` question, it looks like it _should_ get executed when it wakes up http://stackoverflow.com/questions/6346849/what-happens-to-settimeout-when-the-computer-goes-to-sleep

**Comment by Alys** [comment]
@paglias So this change WILL cause cron to run at the CDS time if HabitRPG has been left open in a browser on a non-hibernated computer - is that right?

What about if the HabitRPG app is running on a smartphone overnight? Will that cause cron to run at CDS?

**Comment by paglias** [comment]
@Alys it's right. As for phone we can add it there too but I think the app doesn't run in background

**Comment by Alys** [comment]
@lemoness @blade @deilann and anyone else who often does user support in the Tavern:

This change probably won't affect many people, but we might see occasional questions that result from it, and we should start mentioning it when we describe the time at which cron happens.

Typically, we'll tell a user that cron happens when they take their first action on HabitRPG after their Custom Day Start (CDS) time. This will still be true if they do **not** leave HabitRPG running in a browser on an awake, internet-connected PC. However, if they do leave HabitRPG open and net-connected, then cron will occur at their CDS time.

For most people, this won't make any difference. There are some cases though where it might. If a player is used to having a "necromancer" Healer heal them from negative HP before they cron, then leaving HabitRPG open could make that impossible (or at least reduce the window of time in which the necromancer can save them). Those people will want to be sure to close the HabitRPG website before going to bed.

At the moment, that's the only case I can think of where this cron change will make any difference.

**Comment by lemoness** [comment]
Would it make sense to say something like: "Cron happens at your Custom Day
Start, unless the browser is closed, in which case it will happen as soon
as you next open Habit"?

**Comment by Alys** [comment]
It might do. It sounds almost like the opposite of what we've been telling them so far, but I think it makes more sense.

**Comment by paglias** [comment]
I think that what @Alys outlined about healing is just a trick and the assumption should be that cron runs at cds as @lemoness says

**Comment by Alys** [comment]
_tl;dr: Users understanding cron timing properly is helpful to them and reduces questions. I'll put a suitable description on the wiki and we'll make sure the Tavern regulars understand the new situation._

I don't think we should let it be an assumption that cron runs at CDS. It was for a long time and we've only fairly recently gotten most people realising that cron can run after CDS. My feeling from what I've seen is that many (most?) people don't leave HabitRPG open, so many crons will happen when the users take their first action each day. If users are expecting cron to happen at CDS and it doesn't, then we get repeated questions about why their quest progress is being applied several hours after it "should" be.

Another reason to ensure they understand cron is so they can manage party damage better. If you know you're going to do a lot of damage from incomplete Dailies, but you also know your party mate is going to finish the quest when they cron, then if you have a proper understanding of cron timing, you realise that you can just not use the site for x hours longer than usual and avoid killing your friends. It might not be as good a solution as actually doing your dailies... but the point is that understanding cron accurately empowers the users.

So, yeah, I'll update the wiki and I'll keep explaining the new situation in the Tavern until the new correct wording catches on. :)
